plugin-server/src/main/ingestion-queues/batch-processing/each-batch-ingestion.ts

```diff
@@ -7,19 +7,51 @@ import { sanitizeEvent } from '../../../utils/utils'
 import { KafkaQueue } from '../kafka-queue'
 import { eachBatch } from './each-batch'
 
-export async function eachMessageIngestion(message: KafkaMessage, queue: KafkaQueue): Promise<void> {
+export function getPluginEvent(message: KafkaMessage): PluginEvent {
+    // TODO: inefficient to do this twice?
     const { data: dataStr, ...rawEvent } = JSON.parse(message.value!.toString())
     const combinedEvent = { ...rawEvent, ...JSON.parse(dataStr) }
     const event: PluginEvent = sanitizeEvent({
         ...combinedEvent,
         site_url: combinedEvent.site_url || null,
         ip: combinedEvent.ip || null,
     })
-    await ingestEvent(queue.pluginsServer, queue.workerMethods, event)
+    return event
+}
+
+export async function eachMessageIngestion(message: KafkaMessage, queue: KafkaQueue): Promise<void> {
+    await ingestEvent(queue.pluginsServer, queue.workerMethods, getPluginEvent(message))
 }
 
 export async function eachBatchIngestion(payload: EachBatchPayload, queue: KafkaQueue): Promise<void> {
-    await eachBatch(payload, queue, eachMessageIngestion, 'ingestion')
+    function groupIntoBatchesIngestion(array: KafkaMessage[], batchSize: number): KafkaMessage[][] {
+        // Once we see a distinct ID we've already seen break up the batch
+        const batches = []
+        const seenIds: Set<string> = new Set()
+        let currentBatch: KafkaMessage[] = []
+        for (let i = 0; i < array.length; i += 1) {
+            const message = array[i]
+            const pluginEvent = getPluginEvent(message)
+            const enabledBreakupById = queue.pluginsServer.ingestionBatchBreakupByDistinctIdTeams.has(
```

Comment:
I'm not sure if it makes sense to test this per-team, as that may hide the real performance impact of this solution, and that's the only danger we want to monitor here (as this practically can't affect data integrity negatively).

Reply:
The idea here was gradual rollout because of the potential perf impact (agree it won't impact data integrity):
1. with no teams - does the batching being slower cause problems?
2. enable for a single team - it's going to break the batches more, but given most teams aren't enabled we'll still see rare small batches. If we enable this for the test team first we can also instead of waiting for a deploy stop sending events there or if we want to be more careful start with a team getting no events and gradually ramp up the load for that team.
3. a couple of teams (including the biggest users)
4. everyone

The main slowdown I suspect would come from getting a lot of small batches, but it could also be fine if we're consuming messages from multiple partitions concurrently 🤷‍♀️